src/mapstorage/XmlMapStorage.cpp

```diff
@@ -0,0 +1,420 @@
+};
+static const char *const markClassNames[] = {
+    X_FOREACH_INFOMARK_CLASS(DECL) //
+};
+static const char *const markTypeNames[] = {
+    X_FOREACH_INFOMARK_TYPE(DECL) //
+};
+static const char *const mobFlagNames[] = {
+    X_FOREACH_ROOM_MOB_FLAG(DECL) //
+};
+static const char *const portableNames[] = {
+    X_FOREACH_RoomPortableEnum(DECL) //
+};
+static const char *const ridableNames[] = {
+    X_FOREACH_RoomRidableEnum(DECL) //
+};
+static const char *const sundeathNames[] = {
+    X_FOREACH_RoomSundeathEnum(DECL) //
+};
+static const char *const terrainNames[] = {
+    X_FOREACH_RoomTerrainEnum(DECL) //
+};
+#undef DECL
+
+void XmlMapStorage::saveDoorFlags(QXmlStreamWriter &stream, DoorFlags fl)
+{
+    if (fl.isEmpty()) {
+        return;
+    }
+    QString list;
+    const char *separator = "";
+    for (uint x = 0; x < sizeof(doorFlagNames) / sizeof(doorFlagNames[0]); x++) {
+        if (fl.contains(DoorFlagEnum(x))) {
+            list += separator;
+            separator = ",";
+            list += doorFlagNames[x];
+        }
+    }
+    saveXmlAttribute(stream, "doorflags", list);
+}
+
+void XmlMapStorage::saveExitFlags(QXmlStreamWriter &stream, ExitFlags fl)
+{
+    fl.remove(ExitFlagEnum::EXIT); // always set, do not save it
+    if (fl.isEmpty()) {
+        return;
+    }
+    QString list;
+    const char *separator = "";
+    for (uint x = 0; x < sizeof(exitFlagNames) / sizeof(exitFlagNames[0]); x++) {
+        if (fl.contains(ExitFlagEnum(x))) {
+            list += separator;
+            separator = ",";
+            list += exitFlagNames[x];
+        }
+    }
+    saveXmlAttribute(stream, "flags", list);
+}
+
+void XmlMapStorage::saveRoomLoadFlags(QXmlStreamWriter &stream, RoomLoadFlags fl)
+{
+    if (fl.isEmpty()) {
+        return;
+    }
+    stream.writeStartElement("flags");
+    QString separator;
+    for (uint x = 0; x < sizeof(loadFlagNames) / sizeof(loadFlagNames[0]); x++) {
+        if (fl.contains(RoomLoadFlagEnum(x))) {
+            stream.writeCharacters(separator);
+            if (separator.isEmpty()) {
+                separator = ",";
+            }
+            stream.writeCharacters(loadFlagNames[x]);
+        }
+    }
+    stream.writeEndElement();
+}
+
+void XmlMapStorage::saveRoomMobFlags(QXmlStreamWriter &stream, RoomMobFlags fl)
+{
+    if (fl.isEmpty()) {
+        return;
+    }
+    stream.writeStartElement("mobflags");
+    QString separator;
+    for (uint x = 0; x < sizeof(mobFlagNames) / sizeof(mobFlagNames[0]); x++) {
+        if (fl.contains(RoomMobFlagEnum(x))) {
+            stream.writeCharacters(separator);
+            if (separator.isEmpty()) {
+                separator = ",";
+            }
+            stream.writeCharacters(mobFlagNames[x]);
+        }
+    }
+    stream.writeEndElement();
+}
+
+NODISCARD QString XmlMapStorage::alignName(const RoomAlignEnum x)
+{
+    return enumName(uint(x), alignNames, sizeof(alignNames) / sizeof(alignNames[0]));
```

Comment:
Why even generate alignNames? We could just have that X_FOREACH macro here and return the value

Reply:
Well, I could use a `switch (x)` and a `X_FOREACH_*` macro inside it expanding to each case.
I guess a decent compiler would produce basically the same code.

Reply:
Yeah, plus the macro since X_FOREACH is used to even generate the enum itself so it's extremely safe.

Reply:
Done.